### PR TITLE
Support MongoDB case insensitive collection name

### PR DESCRIPTION
**Pull request by tchunwei**
Fixes https://github.com/prestosql/presto/issues/1102

Currently any MongoDB collection with uppercase name like "Hello" could not be queried.
This PR makes it possible by supporting case insensitivity for collection name, providing solution for case 2 mentioned here https://github.com/prestosql/presto/issues/1102#issuecomment-536331360, it doesn't help case 1 though.

### CODE REVIEW

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoColumnHandle.java [hunk 1]
```diff
@@ -31,18 +33,23 @@
     public static final String SAMPLE_WEIGHT_COLUMN_NAME = "presto_sample_weight";
 
     private final String name;
+    private final String alias;
```
Comment (findepi) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoColumnHandle.java]: do we need a new field?

can we just pass original (un-lower-cased) `name` into `ColumnMetadata#properties`, as it's done in https://github.com/prestosql/presto/pull/1914 ?

File: presto-mongodb/pom.xml [hunk 2]
```diff
@@ -14,8 +14,8 @@
     <properties>
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
         <mongo-java.version>3.6.0</mongo-java.version>
-        <mongo-server.version>1.5.0</mongo-server.version>
-        <netty.version>4.0.32.Final</netty.version>
+        <mongo-server.version>1.22.0</mongo-server.version>
```
Comment (findepi) [inline presto-mongodb/pom.xml]: curious, is mongo version change related to case insensitive collection names?

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoColumnHandle.java (hunk 1 above)
Comment (tchunwei) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoColumnHandle.java]: After reviewing the code yes it is redundant, have removed it.

File: presto-mongodb/pom.xml (hunk 2 above)
Comment (tchunwei) [reply presto-mongodb/pom.xml]: Nope, without updating the version will work also.

**Comment by martint** [comment]
@cla-bot check

File: presto-mongodb/pom.xml (hunk 2 above)
Comment (ebyhr) [reply presto-mongodb/pom.xml]: @tchunwei Let's revert the change of pom.xml so that we can focus on the actual issue. Also, could you resolve conflicts?

Comment (tchunwei) [reply presto-mongodb/pom.xml]: have reverted the version and pull latest master in

**Comment by findepi** [comment]
@tchunwei could you please rebase on current master and squash commits? Ideally, there should be just one commit after that.

File: presto-mongodb/pom.xml (hunk 2 above)
Comment (ebyhr) [reply presto-mongodb/pom.xml]: Thanks, please revert the change for `org.slf4j` too.

Comment (tchunwei) [reply presto-mongodb/pom.xml]: @ebyhr sorry I missed that out, has reverted it too.

**Comment by tchunwei** [comment]
@findepi thanks for guiding, has rebased, made it into single commit and force pushed. Kindly let me know if I did it incorrectly, thanks again.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java [hunk 3]
```diff
@@ -88,9 +87,14 @@ public MongoTableHandle getTableHandle(ConnectorSession session, SchemaTableName
     @Override
     public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        requireNonNull(tableHandle, "tableHandle is null");
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java]: This change isn't related to this commit.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java [hunk 4]
```diff
@@ -101,7 +105,7 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
         ImmutableList.Builder<SchemaTableName> tableNames = ImmutableList.builder();
         for (String schemaName : schemaNames) {
             for (String tableName : mongoSession.getAllTables(schemaName)) {
-                tableNames.add(new SchemaTableName(schemaName, tableName.toLowerCase(ENGLISH)));
+                tableNames.add(new SchemaTableName(schemaName, tableName));
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java]: While the `toLowerCase` is redundant, I don't think this change is required for this PR. I would suggest separating a commit.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java [hunk 5]
```diff
@@ -153,34 +157,36 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+        mongoSession.createTable(table, buildColumnHandles(tableMetadata));
     }
 
     @Override
     public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
         MongoTableHandle table = (MongoTableHandle) tableHandle;
-
-        mongoSession.dropTable(table.getSchemaTableName());
+        mongoSession.dropTable(table);
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java]: Revert the change for removing the empty line.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java [hunk 6]
```diff
@@ -153,34 +157,36 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
     }
 
     @Override
     public void renameTable(ConnectorSession session, ConnectorTableHandle tableHandle, SchemaTableName newTableName)
     {
-        throw new UnsupportedOperationException();
+        MongoTableHandle table = (MongoTableHandle) tableHandle;
+        mongoSession.renameTable(table, newTableName);
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoMetadata.java]: Revert this change because it's unrelated to the case insensitive commit. I would recommend sending another PR with the test.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoPageSink.java [hunk 7]
```diff
@@ -76,36 +75,36 @@
...
+        MongoCollection<Document> collection = mongoSession.getCollection(table);
         List<Document> batch = new ArrayList<>(page.getPositionCount());
 
         for (int position = 0; position < page.getPositionCount(); position++) {
             Document doc = new Document();
 
             for (int channel = 0; channel < page.getChannelCount(); channel++) {
                 MongoColumnHandle column = columns.get(channel);
-                doc.append(column.getName(), getObjectValue(columns.get(channel).getType(), page.getBlock(channel), position));
+                doc.append(column.getName(), getObjectValue(column.getType(), page.getBlock(channel), position));
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoPageSink.java]: I suppose that this change isn't mandatory to support case insensitive collection name. Please let me know if I am wrong.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 8]
```diff
@@ -120,73 +123,101 @@ public MongoSession(TypeManager typeManager, MongoClient client, MongoClientConf
...
     }
 
     public void shutdown()
     {
         client.close();
     }
 
+    /**
+     * @return lower-case schema names
+     */
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java]: Remove this comment to be consistent with other methods (getAllTables is also the same)

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 9]
```diff
@@ -120,73 +123,101 @@ public MongoSession(TypeManager typeManager, MongoClient client, MongoClientConf
...
 
-        ImmutableList.Builder<MongoColumnHandle> columnHandles = ImmutableList.builder();
+        tableCache.invalidate(table.getSchemaTableName());
+    }
 
-        for (Document columnMetadata : getColumnMetadata(tableMeta)) {
-            MongoColumnHandle columnHandle = buildColumnHandle(columnMetadata);
-            columnHandles.add(columnHandle);
-        }
+    public void renameTable(MongoTableHandle table, SchemaTableName newTableName)
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java]: Revert this change.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 10]
```diff
@@ -438,37 +454,25 @@ private void createTableMetadata(SchemaTableName schemaTableName, List<MongoColu
...
-        String schemaName = schemaTableName.getSchemaName();
-        String tableName = schemaTableName.getTableName();
-
-        MongoDatabase db = client.getDatabase(schemaName);
-        Document doc = db.getCollection(tableName).find().first();
+        Document doc = getCollection(table).find().first();
         if (doc == null) {
             // no records at the collection
-            return ImmutableList.of();
+            throw new PrestoException(NOT_SUPPORTED, "Empty collection");
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java]: Revert this change.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 11]
```diff
@@ -558,4 +562,60 @@ else if (value instanceof Document) {
 
         return Optional.ofNullable(typeSignature);
     }
+
+    private String getDatabaseName(String lname)
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java]: We don't use an abbreviation. Let's rename `lname` and the following `a`, `lDatabaseName` and `lTableName`

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoTableHandle.java [hunk 12]
```diff
@@ -27,27 +27,41 @@
...
     {
-        this(schemaTableName, TupleDomain.all());
+        this(table.getSchemaName(), table.getTableName(), TupleDomain.all());
     }
 
     @JsonCreator
-    public MongoTableHandle(
-            @JsonProperty("schemaTableName") SchemaTableName schemaTableName,
-            @JsonProperty("constraint") TupleDomain<ColumnHandle> constraint)
+    public MongoTableHandle(@JsonProperty("databaseName") String databaseName,
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoTableHandle.java]: `@JsonProperty("databaseName") String databaseName,` should be next line

File: presto-mongodb/src/test/java/io/prestosql/plugin/mongodb/TestMongoIntegrationSmokeTest.java [hunk 13]
```diff
@@ -285,6 +285,16 @@ public void testObjectIds()
         assertUpdate("DROP TABLE tmp_objectid");
     }
 
+    @Test
+    public void testCamelCase()
```
Comment (ebyhr) [inline presto-mongodb/src/test/java/io/prestosql/plugin/mongodb/TestMongoIntegrationSmokeTest.java]: Let's rename to `testCaseInsensitive`.

**Comment by buffcode** [comment]
@tchunwei  Why was this closed?

**Comment by tchunwei** [comment]
@buffcode that was done unintentionally, was rebasing my code then accidentally wiped my commit, will re-open later

**Comment by tchunwei** [comment]
Thank you @ebyhr for your effort guiding me and reviewing the code. Have updated the code based on the feedback for your further review.

**Comment by zifer123** [comment]
any updated about this?

**Comment by tchunwei** [comment]
I am sorry I do not have much time for this and this is consider low priority for me so will put this aside first. Also, this PR will be a temporary solution until https://github.com/prestosql/presto/pull/2350 is ready. So I think it would be better to wait for https://github.com/prestosql/presto/pull/2350 since it is a better approach?

**Comment by zifer123** [comment]
> I am sorry I do not have much time for this and this is consider low priority for me so will put this aside first. Also, this PR will be a temporary solution until #2350 is ready. So I think it would be better to wait for #2350 since it is a better approach?

ok, thanks

**Comment by d4rth-v4d3r** [comment]
Any updates of this?

**Comment by ebyhr** [comment]
Covered by #3453
